### PR TITLE
add all configuration options

### PR DESCRIPTION
**Pull request by bobzilladev**
All configuration options for all tunnel types are now settable, and the data flows onto the wire. Have manually tested everything to work ~~except for Mutual TLS, TLS cert/key options, and a functionaing OIDC~~. Will check those next, but this diff is large so wanted to put it up for review in parallel.

Edit: Got OIDC working with an Okta trial account, and TLS with a self-signed cert, and Mutual TLS with self signed chain.

Some questions were embedded (now answered):
- oauth_options // how to chain 'new' to 'with_*'?
- // todo: ngrok-rs has "hostname" and "subdomain", ngrok-go has just "domain"?
- pub(crate) mutual_tlsca: Vec<Vec<u8>>, // something more idiomatic here?
- pub(crate) key_pem: Option<Vec<u8>>, // something more idiomatic here?
- pub(crate) cert_pem: Option<Vec<u8>>, // something more idiomatic here?

### CODE REVIEW

**Comment by CompeyDev** [comment]
Hey @jrobsonchase, can this library even forward local ports to a tunnel? I spent hours going through docs and still could not find any methods to set the local port of the tunnel; would really appreciate if you could link me a doc or mention a way to do this.

**Comment by russorat** [comment]
Hi @CompeyDev, thanks for commenting! This library is built directly into your rust application so there's no ports to forward. Traffic goes directly to your rust code.

I invited you to our Slack community (https://ngrok.com/slack) if you want to chat about this more.

**Comment by CompeyDev** [comment]
Ah, I thought this was an official wrapper of sorts, similar to a few other crates. 

> I invited you to our Slack community (https://ngrok.com/slack) if you want to chat about this more.

That's amazing, but I unfortunately do not use Slack, apologies!

**Comment by jrobsonchase** [comment]
@CompeyDev Yep, what @russorat said!

If you want to see some example usage, https://github.com/ngrok/ngrok-rs/blob/main/ngrok/examples/ should give you a place to start. 

TL;DR: you can use a `Tunnel` like a TCP listener. It implements the async `Stream<Item = Result<Conn, Error>>` trait as well as `hyper`'s `server::Accept` trait so you can use it directly as the listener backend to any `hyper`-based web servers.

**Comment by CompeyDev** [comment]
Yeah, I get that -- I think I just totally misunderstood the purpose of the library; seeing the less number of examples, I just thought the examples weren't fully completed and documented. Thanks for the clarification, though!
